cw_bitcoin/lib/electrum_wallet_addresses.dart

```diff
@@ -187,126 +233,138 @@ abstract class ElectrumWalletAddressesBase extends WalletAddresses with Store {
     if (address.startsWith('bitcoincash:')) {
       address = toLegacy(address);
     }
-    final addressRecord = addresses.firstWhere((addressRecord) => addressRecord.address == address);
+    final addressRecord =
+        _addresses.firstWhere((addressRecord) => addressRecord.address == address);
     addressRecord.setNewName(label);
-    final index = addresses.indexOf(addressRecord);
-    addresses.remove(addressRecord);
-    addresses.insert(index, addressRecord);
+    final index = _addresses.indexOf(addressRecord);
+    _addresses.remove(addressRecord);
+    _addresses.insert(index, addressRecord);
+  }
+
+  @action
+  void updateAddressesByMatch() {
+    addresses.clear();
+    addresses.addAll(_addresses.where(_isAddressPageTypeMatch).toList());
   }
 
   @action
   void updateReceiveAddresses() {
     receiveAddresses.removeRange(0, receiveAddresses.length);
     final newAddresses =
-        addresses.where((addressRecord) => !addressRecord.isHidden && !addressRecord.isUsed);
+        _addresses.where((addressRecord) => !addressRecord.isHidden && !addressRecord.isUsed);
     receiveAddresses.addAll(newAddresses);
   }
 
   @action
   void updateChangeAddresses() {
     changeAddresses.removeRange(0, changeAddresses.length);
-    final newAddresses =
-        addresses.where((addressRecord) => addressRecord.isHidden && !addressRecord.isUsed);
+    final newAddresses = _addresses.where((addressRecord) =>
+        addressRecord.isHidden &&
+        !addressRecord.isUsed &&
+        // TODO: feature to change change address type. For now fixed to p2wpkh, the cheapest type
+        addressRecord.type == SegwitAddresType.p2wpkh);
     changeAddresses.addAll(newAddresses);
   }
 
-  Future<void> _discoverAddresses(bitcoin.HDWallet hd, bool isHidden) async {
-    var hasAddrUse = true;
-    List<BitcoinAddressRecord> addrs;
-
-    if (addresses.isNotEmpty) {
-
-
-      if(!isHidden) {
-        final receiveAddressesList = addresses.where((addr) => !addr.isHidden).toList();
-        validateSideHdAddresses(receiveAddressesList);
-      }
-
-      addrs = addresses.where((addr) => addr.isHidden == isHidden).toList();
-    } else {
-      addrs = await _createNewAddresses(
-          isHidden ? defaultChangeAddressesCount : defaultReceiveAddressesCount,
-          startIndex: 0,
-          hd: hd,
-          isHidden: isHidden);
+  @action
+  Future<void> discoverAddresses(List<BitcoinAddressRecord> addressList, bool isHidden,
```

Comment:
I don't see you calling this with the change address list or even with isHidden set to true